edgelet/aziot-edged/src/watchdog.rs

```diff
@@ -111,16 +115,74 @@ async fn watchdog(
                     .await
                     .map_err(|err| EdgedError::from_err("Failed to remove Edge runtime", err))?;
 
-                create_and_start_agent(settings, device_info, runtime, identity_client).await?;
+                create_and_start_agent(settings, device_info, &runtime, identity_client).await?;
             }
         }
     } else {
-        create_and_start_agent(settings, device_info, runtime, identity_client).await?;
+        create_and_start_agent(settings, device_info, &runtime, identity_client).await?;
     }
 
     Ok(())
 }
 
+async fn restart_modules(
+    settings: &edgelet_settings::docker::Settings,
+    runtime: &edgelet_docker::DockerModuleRuntime,
+) {
+    let agent_name = settings.agent().name();
+
+    // Check if edgeAgent is running. If edgeAgent does not exist or is not running,
+    // return and let the periodic watchdog create and start it.
+    if let Ok((_, agent_status)) = runtime.get(agent_name).await {
+        if agent_status.status() != &edgelet_core::ModuleStatus::Running {
+            log::info!("Agent not running; skipping module restart");
+
+            return;
+        }
+    } else {
+        log::info!("Agent not found; skipping module restart");
+
+        return;
+    }
+
+    // List and stop modules.
+    let modules = if let Ok(modules) = runtime.list().await {
+        modules
+    } else {
+        log::warn!("Failed to list modules");
+
+        return;
+    };
+
+    if let Err(err) = runtime.stop_all(None).await {
+        log::warn!("Edge CA renewal failed to stop modules: {}", err);
+
+        return;
+    }
+
+    log::info!("Edge CA renewal stopped all modules");
+
+    // Restart all modules. edgeAgent should be restarted last so that it does not
```

Comment:
Won't it be better to only start EA and nothing else, and let EA start the other modules. Eg if a new deployment happens at the same time as this renewal, you might end up restarting a module that needn't have been restarted.

Reply:
Yes, good point.

The race isn't a problem anyway, because even if we do start a superfluous module it will be tagged such that EA will stop it in its next reconcile loop iteration.